magma/debug.py

```diff
@@ -1,43 +1,31 @@
-import inspect
+import uinspect
 import collections
-import magma
-from magma.config import get_debug_mode
 
 
-debug_info = collections.namedtuple("debug_info", ["filename", "lineno", "module"])
+debug_info = collections.namedtuple("debug_info",
+                                    ["filename", "lineno", "module"])
 
 
-def get_callee_frame_info():
-    callee_frame = inspect.currentframe()
-    # FIXME: Right now we assume a max 10 frames deep
-    for i in range(0, 10):
-        module = inspect.getmodule(callee_frame)
-        # Go up until we're out of the magma module (assuming this is the user
-        # code)
-        if not module or module.__name__.split(".")[0] != "magma":
-            break
-        callee_frame = callee_frame.f_back
-    callee_frame = inspect.getframeinfo(callee_frame)
-    return debug_info(callee_frame.filename, callee_frame.lineno, module)
+def get_debug_info(frames_to_skip):
```

Comment:
do we always know `frames_to_skip` statically? vs. the popping up dynamically...I kind of like the dynamic one, unless this is pretty robust.

Reply:
The dynamic one requires traversing the stack with the inspect module which I think is one of the reasons why the current debug mode is so slow.

One option would be to have the C api do the traversal out of the magma module for us, but that doesn't quite work well for one reason: the primitives are inside the magma module and we'd want debug info to be generated there.  I think in general it's faster and easier to specify frames_to_skip the in places where we know we want to skip frames

Reply:
> the primitives are inside the magma module and we'd want debug info to be generated there.

If you can specify a list of files to allow and discard, I can introduce a C API that traverse the stack quickly. I believe it should be orders of magnitude faster than `inspect` module.

Reply:
We could try that by passing the list of core magma .py files (so basically traverse outside of the core codebase to try to figure out where the wire occured)

Reply:
@leonardt Can you try the latest `uinspect`?  I've added a simple frame walking logic that look for any frames that don't originate from the block list (see [here](https://github.com/Kuree/uinspect/blob/c18449a03dd200da35ecd4f5abee3f5a1446cea3/tests/test_uinspect.py#L51)).

Based on the benchmark, it's about 35x faster than walking the stack in Python, about 1.5us/op. So it's still 5x more expensive than static methods.

Reply:
Could we try adding an API to register the files to skip, then invoke the stack walking without having to pass the file names each time? (Perhaps avoiding passing the strings each time would speed things up a bit)

Reply:
You only need to create the `FrameWalker` object once, by providing a set of strings. Then any subsequent calls of `get_location()` will walk the stack. Here is some usage:

```python
walker = uinspect.FrameWalker({"foo.py"})


def foo():
   loc = walker.get_location()


def bar():
  loc = walker.get_location()

```


I think the slowdown is caused by casting Python string into std::string, and then search in an unordered_set. If a more efficient string hash data structure is used, I think we can speed it up more. Let me do more experiments.

Reply:
do we have a conclusion here?

Reply:
I think we should just use static frames to skip for now and add this support as an improvement

Reply:
let's make an issue for this to track then.